views/student.py

```diff
@@ -280,54 +280,71 @@ def student_save_affectation_modification(request, cohort_id, student_id):
     if 'internship' in request.POST:
         internships = request.POST.getlist('internship')
 
-    check_error_present = False
-    num_periods = len(periods)
-    for num_period in range(0, num_periods):
-        if organizations[num_period]:
-            organization = mdl_int.organization.search(cohort=cohort, reference=organizations[num_period])[0]
-            specialty = mdl_int.internship_speciality.search(cohort=cohort, name=specialties[num_period])[0]
-            period = mdl_int.period.search(cohort=cohort, name=periods[num_period])[0]
-            check_internship_present = mdl_int.internship_offer.search(cohort=cohort,
-                                                                       organization=organization,
-                                                                       speciality=specialty)
-            if len(check_internship_present) == 0:
-                check_error_present = True
-                messages.add_message(request, messages.ERROR, "{} : {}-{} ({})=> error".format(specialty.name,
-                                                                                               organization.reference,
-                                                                                               organization.name,
-                                                                                               period.name))
-
-    student_affectations = InternshipStudentAffectationStat.objects.filter(student=student, period__cohort=cohort)
-    if _has_validated_score(student_affectations):
-        check_error_present = True
+    update_data = _build_update_data(cohort, organizations, specialties, internships)
+    error = _check_error_present(cohort, periods, request, update_data)
+
+    student_affectations_to_update = InternshipStudentAffectationStat.objects.filter(
+        student=student,
+        period__cohort=cohort,
+        period__name__in=periods
+    ).order_by('period__date_start')
+
+    if _has_validated_score(student_affectations_to_update):
```

Comment:
if _check_error_present(...) or _has_validated_score(..):
        messages.add_message(
            request, messages.ERROR, _(
                "Cannot edit affectations because at least one affectation has a linked validated score"
            )
        )
        redirect_url = ...
else:
        update_selected_student_affectations(...)
        redirect_url = ...